### PR TITLE
Support for controlSize(:) and frame(width:) modifier

### PR DESCRIPTION
**Pull request by Maschina**
My current app design requires a smaller variant of the text field. Thus, I have implemented a proper support for controlSize(:) that (among others) reflects the font size to the controlSize parameter. Furthermore, changing the text field's width was an important requirements for proper layouting.

FYI: Please refer to the Previews in Recorder.swift for testings.

### CODE REVIEW

**Comment by sindresorhus** [comment]
I would prefer a solution that doesn't require introducing a new option.

File: Sources/KeyboardShortcuts/Recorder.swift [hunk 1]
```diff
@@ -31,39 +31,70 @@ extension KeyboardShortcuts {
 
 		private let name: Name
 		private let onChange: ((_ shortcut: Shortcut?) -> Void)?
+        private let isHorizontalContentSizeConstraintActive: Bool
```
Comment (sindresorhus) [inline Sources/KeyboardShortcuts/Recorder.swift:34]: Even if we decide to introduce a new option, this option is too specific.

File: Sources/KeyboardShortcuts/Recorder.swift [hunk 2]
```diff
@@ -31,39 +31,70 @@ extension KeyboardShortcuts {
...
 
 		/// :nodoc:
 		public func updateNSView(_ nsView: NSViewType, context: Context) {
 			nsView.shortcutName = name
+            
+            // support for controlSize(:) modifier
+            switch context.environment.controlSize {
+                case .small:
+                    nsView.font = .systemFont(ofSize: 11)
+                    nsView.changeWidth(width: 110)
```
Comment (sindresorhus) [inline Sources/KeyboardShortcuts/Recorder.swift:62]: Why manually apply this instead of just using `NSControl#controlSize`?

**Comment by Maschina** [comment]
> I would prefer a solution that doesn't require introducing a new option.

You are referring to `isHorizontalContentSizeConstraintActive`?

File: Sources/KeyboardShortcuts/Recorder.swift [hunk 3]
```diff
@@ -31,39 +31,70 @@ extension KeyboardShortcuts {
...
 			KeyboardShortcuts.Recorder(for: .init("xcodePreview"))
 				.environment(\.locale, .init(identifier: "en"))
-			KeyboardShortcuts.Recorder(for: .init("xcodePreview"))
-				.environment(\.locale, .init(identifier: "zh-Hans"))
-			KeyboardShortcuts.Recorder(for: .init("xcodePreview"))
-				.environment(\.locale, .init(identifier: "ru"))
+                .frame(width: 130)
+            
+            KeyboardShortcuts.Recorder(for: .init("xcodePreview"))
+                .previewDisplayName("Test: .controlSize(.small)")
```
Comment (sindresorhus) [inline Sources/KeyboardShortcuts/Recorder.swift:82]: Please use tab-indentation everywhere and make sure listing passes.

File: Sources/KeyboardShortcuts/RecorderCocoa.swift [hunk 4]
```diff
@@ -67,10 +67,10 @@ extension KeyboardShortcuts {
...
 		}
 
-		/**
-		- Parameter name: Strongly-typed keyboard shortcut name.
-		- Parameter onChange: Callback which will be called when the keyboard shortcut is changed/removed by the user. This can be useful when you need more control. For example, when migrating from a different keyboard shortcut solution and you need to store the keyboard shortcut somewhere yourself instead of relying on the built-in storage. However, it's strongly recommended to just rely on the built-in storage when possible.
-		*/
+        /**
+         - Parameter name: Strongly-typed keyboard shortcut name.
+         - Parameter onChange: Callback which will be called when the keyboard shortcut is changed/removed by the user. This can be useful when you need more control. For example, when migrating from a different keyboard shortcut solution and you need to store the keyboard shortcut somewhere yourself instead of relying on the built-in storage. However, it's strongly recommended to just rely on the built-in storage when possible.
+         */
```
Comment (sindresorhus) [inline Sources/KeyboardShortcuts/RecorderCocoa.swift:73]: Don't do unrelated changes.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: Sources/KeyboardShortcuts/Recorder.swift (hunk 1 above)
Comment (Maschina) [reply Sources/KeyboardShortcuts/Recorder.swift:34]: Any counter proposal? The actual init parameter is called `constraint`.

File: Sources/KeyboardShortcuts/Recorder.swift (hunk 2 above)
Comment (Maschina) [reply Sources/KeyboardShortcuts/Recorder.swift:62]: Good point, will fix it

**Comment by sindresorhus** [comment]
> You are referring to isHorizontalContentSizeConstraintActive?

Yes

File: Sources/KeyboardShortcuts/Recorder.swift (hunk 1 above)
Comment (Maschina) [reply Sources/KeyboardShortcuts/Recorder.swift:34]: We might go this approach: We completely get right of the built-in width constraint. The control is utilizing its width to the available space completely. 

Seeing how all other native SwiftUI controls are implemented, I would say that this behavior would be more SwiftUI'ful, because their widths is always controlled through either modifiers (e.g. `.frame()`), or ViewBuilder structs.
I haven't seen any native SwiftUI control that has a fixed width constraint.

What do you think?

Comment (sindresorhus) [reply Sources/KeyboardShortcuts/Recorder.swift:34]: I partly agree. The control still needs a minimum width (and intrinsic content size) to make sure the label is not cut off.

Comment (Maschina) [reply Sources/KeyboardShortcuts/Recorder.swift:34]: OK, I see your point. Nevertheless, this is a topic every UI designer has to handle with each and every other native controls such as password textfield etc. 

I see your point that the keyboard recorder control requires a minimum width.
Maybe we could control the minimum width in the `override public var intrinsicContentSize: CGSize { ... }` function by not allowing a width of the size below a certain value?

Comment (sindresorhus) [reply Sources/KeyboardShortcuts/Recorder.swift:34]: > I agree with you that the keyboard recorder control requires a minimum width.
Maybe we could control the minimum width in the override public var intrinsicContentSize: CGSize { ... } function by not allowing a width of the size below a certain value?

👍

File: Sources/KeyboardShortcuts/RecorderCocoa.swift [hunk 5]
```diff
@@ -88,9 +92,9 @@ extension KeyboardShortcuts {
 			self.wantsLayer = true
 			self.translatesAutoresizingMaskIntoConstraints = false
 			setContentHuggingPriority(.defaultHigh, for: .vertical)
-			setContentHuggingPriority(.defaultHigh, for: .horizontal)
-			widthAnchor.constraint(greaterThanOrEqualToConstant: CGFloat(minimumWidth)).isActive = true
-
+			setContentHuggingPriority(.defaultLow, for: .horizontal)
+			self.isHorizontalContentSizeConstraintActive = true
```
[the remark this answers is not in the document]
Comment (sindresorhus) [reply Sources/KeyboardShortcuts/RecorderCocoa.swift:96]: It also requires macOS 10.15, so it needs to be guarded.

File: Sources/KeyboardShortcuts/RecorderCocoa.swift [hunk 6]
```diff
@@ -138,6 +142,15 @@ extension KeyboardShortcuts {
 				focus()
 			}
 		}
+		
+		public func changeWidth(for size: NSControl.ControlSize) {
```
Comment (sindresorhus) [inline Sources/KeyboardShortcuts/RecorderCocoa.swift:146]: This should not be a public method. The control size should be handled like any other control. I guess, you could override `.controlSize` and handle the changes there.

File: Sources/KeyboardShortcuts/RecorderCocoa.swift [hunk 7]
```diff
@@ -26,10 +26,14 @@ extension KeyboardShortcuts {
...
 	public final class RecorderCocoa: NSSearchField, NSSearchFieldDelegate {
-		private let minimumWidth: Double = 130
 		private var eventMonitor: LocalEventMonitor?
 		private let onChange: ((_ shortcut: Shortcut?) -> Void)?
 		private var observer: NSObjectProtocol?
+		private var intrinsicWidth: CGFloat = 130
+		
+		private let intrinsicWidthRegular: CGFloat = 130
+		private let intrinsicWidthSmall: CGFloat = 110
+		private let intrinsicWidthMini: CGFloat = 95
```
Comment (sindresorhus) [inline Sources/KeyboardShortcuts/RecorderCocoa.swift:36]: We should not hard-code the minimum widths, since we support localization. They should be based on the intrinsic content size of the labels in the button, with some tiny padding on each side.

File: Sources/KeyboardShortcuts/RecorderCocoa.swift [hunk 8]
```diff
@@ -26,10 +26,14 @@ extension KeyboardShortcuts {
 	```
 	*/
 	public final class RecorderCocoa: NSSearchField, NSSearchFieldDelegate {
-		private let minimumWidth: Double = 130
 		private var eventMonitor: LocalEventMonitor?
 		private let onChange: ((_ shortcut: Shortcut?) -> Void)?
 		private var observer: NSObjectProtocol?
+		private var intrinsicWidth: CGFloat = 130
```
Comment (sindresorhus) [inline Sources/KeyboardShortcuts/RecorderCocoa.swift:32]: This is moot. Just use `intrinsicWidthRegular`.

File: Sources/KeyboardShortcuts/RecorderCocoa.swift [hunk 9]
```diff
@@ -68,9 +72,9 @@ extension KeyboardShortcuts {
 		}
 
 		/**
-		- Parameter name: Strongly-typed keyboard shortcut name.
-		- Parameter onChange: Callback which will be called when the keyboard shortcut is changed/removed by the user. This can be useful when you need more control. For example, when migrating from a different keyboard shortcut solution and you need to store the keyboard shortcut somewhere yourself instead of relying on the built-in storage. However, it's strongly recommended to just rely on the built-in storage when possible.
-		*/
+		 - Parameter name: Strongly-typed keyboard shortcut name.
+		 - Parameter onChange: Callback which will be called when the keyboard shortcut is changed/removed by the user. This can be useful when you need more control. For example, when migrating from a different keyboard shortcut solution and you need to store the keyboard shortcut somewhere yourself instead of relying on the built-in storage. However, it's strongly recommended to just rely on the built-in storage when possible.
+		 */
```
Comment (sindresorhus) [inline Sources/KeyboardShortcuts/RecorderCocoa.swift:77]: Always look over the diff after doing changes.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
You still have some lint warnings. Look at the diff.

**Comment by sindresorhus** [comment]
You also need to update the example project and docs to use `.fixedSize()` to make it the intrinsic content size as that's what most users will want.

**Comment by sindresorhus** [comment]
Closing for lack of activity.
